mongodb/mongo_common.cc

```diff
@@ -0,0 +1,216 @@
+        }
+        else if (el.type() == bsoncxx::type::k_array) {
+            row.emplace_back(i, extract(ts, el.get_array()));
+        }
+        else {
+            row.emplace_back(
+                i, ExpressionValue(bsonToCell(el.get_value()), ts));
+        }
+    }
+    return row;
+}
+
+void validateConnectionScheme(const std::string & connectionScheme)
+{
+    // Mongo driver is very sensitive regarding uris.
+    // Bad ones may provoke aborts.
+    if (connectionScheme.empty() || connectionScheme.find("mongodb://") != 0
+            || connectionScheme.size() < 11)
+    {
+        throw HttpReturnException(
+            400,
+            "connectionScheme is a required property and the minimal "
+            "connectionScheme format is mongodb://<host>");
+    }
+}
+
+void validateCollection(const std::string & collection)
+{
+    if (collection.empty()) {
+        throw HttpReturnException(
+            400, "collection is a required property and must not be empty");
+    }
+}
+
+ColumnGetter
+MongoScope::
+doGetColumn(const Utf8String & tableName,
+            const ColumnName & columnName)
+{
+    return {[=] (const SqlRowScope & scope, ExpressionValue & storage,
+                 const VariableFilter & filter) -> const ExpressionValue &
+        {
+            const auto & row = scope.as<MongoRowScope>();
+            const ExpressionValue * res =
+                row.expr.tryGetNestedColumn(columnName, storage, filter);
+            if (res) {
+                return *res;
+            }
+            return storage = ExpressionValue();
+        },
+        std::make_shared<AtomValueInfo>()
+    };
+}
+
+GetAllColumnsOutput
+MongoScope::
+doGetAllColumns(const Utf8String & tableName, const ColumnFilter & keep)
+{
+    std::vector<KnownColumn> columnsWithInfo;
+
+    auto exec = [=] (const SqlRowScope & scope, const VariableFilter & filter)
+    {
+        const auto & row = scope.as<MongoRowScope>();
+        StructValue result;
+        result.reserve(row.expr.rowLength());
+
+        const auto onCol = [&] (const PathElement & columnName,
+                                const ExpressionValue & val)
+        {
+            const auto & newColName = keep(columnName);
+            if (!newColName.empty()) {
+                result.emplace_back(newColName.front(), val);
+            }
+            return true;
+        };
+        row.expr.forEachColumnDestructive(onCol);
+        result.shrink_to_fit();
+        return result;
+    };
+    GetAllColumnsOutput result;
+    result.exec = exec;
+    result.info = std::make_shared<RowValueInfo>(std::move(columnsWithInfo),
+                                                 SCHEMA_OPEN);
+    return result;
+}
+
+BoundFunction
+MongoScope::
+doGetFunction(const Utf8String & tableName,
+              const Utf8String & functionName,
+              const std::vector<BoundSqlExpression> & args,
+              SqlBindingScope & argScope)
+{
+    if (functionName == "oid") {
```

Comment:
This is a special function name for Mongo.  I did not see it documented anywhere.